project-setup

EntyR opened the pull request.

Commit:
Added domain layer in which created randomnumber model repository interface and 2 use cases

AlexeyKolosov commented on domain/src/main/java/com/harman/domain/repository/RandomNumberRepository.kt:
```diff
@@ -0,0 +1,4 @@
+package com.harman.domain.repository
+
+class RundomNumberRepository {
```
Please put this class into data layer, and add an interface for this class here (in domain layer)

AlexeyKolosov commented on domain/src/main/java/com/harman/domain/model/RandomNumber.kt:
```diff
@@ -0,0 +1,3 @@
+package com.harman.domain.model
```
now it seems a little bit strange, but lets keep domain, data and app modules in your `package="com.harman.roomdbapp"`, so this will be like: `com.harman.roomdbapp.domain.model` and please change another modules too.

EntyR replied:
@AlexeyKolosov, you mean to just change package name in modules?

AlexeyKolosov replied:
Yes, seemed that directory was wrong, for now it is good for domain module, please check other modules also

AlexeyKolosov commented:
also please check [this](https://insert-koin.io/docs/quickstart/android#writing-the-koin-module) documentation and try to add Koin

EntyR committed:
included koin di, changed modules package name, changed repository in domain to interface and added repository implementation class in data

EntyR committed:
included room database interfaces in data module

AlexeyKolosov commented on domain/src/main/java/com/harman/roomdbapp/domain/repository/RandomNumberRepository.kt:
```diff
@@ -0,0 +1,12 @@
+import androidx.lifecycle.LiveData
+import com.harman.roomdbapp.domain.model.RandomNumber
+
+interface RandomNumberRepository {
```
Lets name interfaces starting with "I" letter, for example: `interface IRandomNumberRepository`, and in this case we can remove "Imp" from its implementation class.

AlexeyKolosov commented on data/src/main/java/com/harman/roomdbapp/repository/RandomNumberRepositoryImp.kt:
```diff
@@ -0,0 +1,26 @@
+        db.getDao().addNumber(number.convertToEntity())
+    }
+
+    override suspend fun getNumbers(): LiveData<List<RandomNumber>> {
```
as it returns observable-like object, there is no need to mark it with "suspend", please remove "suspend" here. Also, LiveData is observed in the main thread, but this method can also be observed in the background, in this case we can use `Flow`. Please replace here LiveData to Flow. Useful links: [https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-flow/](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-flow/), [https://developer.android.com/kotlin/flow](https://developer.android.com/kotlin/flow)

AlexeyKolosov commented on data/src/main/java/com/harman/roomdbapp/dao/RandomNumberDao.kt:
```diff
@@ -0,0 +1,18 @@
+import com.harman.roomdbapp.enity.RandomNumberEntity
+
+@Dao
+interface RandomNumberDao {
```
lets rename to `IRandomNumberDao`

AlexeyKolosov commented on data/src/main/java/com/harman/roomdbapp/dao/RandomNumberDao.kt:
```diff
@@ -0,0 +1,18 @@
+package com.harman.roomdbapp.dao
```
strange directory, maybe it should be `com.harman.roomdbapp.data.dao` ?

AlexeyKolosov commented on data/src/main/java/com/harman/roomdbapp/EntityMappers.kt:
```diff
@@ -0,0 +1,15 @@
+package com.harman.roomdbapp
```
strange directory, maybe it should be `com.harman.roomdbapp.data` ?

AlexeyKolosov commented on app/src/main/java/com/harman/roomdbapp/di/AppModule.kt:
```diff
@@ -0,0 +1,7 @@
+package com.harman.roomdbapp.di
```
strange directory, maybe it should be `com.harman.roomdbapp.app.di` ?

AlexeyKolosov commented on domain/src/main/java/com/harman/roomdbapp/domain/repository/RandomNumberRepository.kt:
```diff
@@ -0,0 +1,12 @@
+
+    suspend fun addNumber(number: RandomNumber)
+
+    suspend fun getNumbers(): LiveData<List<RandomNumber>>
```
as described above, please remove `suspend` from methods, which return observable-like object. Also here please replace `LiveData` to `Flow`

AlexeyKolosov commented on domain/src/main/java/com/harman/roomdbapp/domain/use_cases/AddNewNumberUseCase.kt:
```diff
@@ -0,0 +1,12 @@
+
+class AddNewNumberUseCase(private val repository: RandomNumberRepository) {
+
+    suspend operator fun invoke(number: RandomNumber){
```
I think, there is no need to use operator keyword here, lets make it just `fun`. Plus, as for the architecture, maybe it is better to have here something like "RandomNumberUseCase", so all your methods (add, getAll, get) can be right in this usecase

AlexeyKolosov commented on domain/src/main/java/com/harman/roomdbapp/domain/use_cases/GetAllNumbersUseCase.kt:
```diff
@@ -0,0 +1,13 @@
+import com.harman.roomdbapp.domain.model.RandomNumber
+import com.harman.roomdbapp.domain.repository.RandomNumberRepository
+
+class GetAllNumbersUseCase(private val repository: RandomNumberRepository) {
```
the same here, see comment above

AlexeyKolosov commented on app/src/main/java/com/harman/roomdbapp/di/AppModule.kt:
```diff
@@ -0,0 +1,7 @@
+import org.koin.dsl.module
+
+val appModule = module {
+
```
As you have already created some use cases and repositories, you can declare them here like: 
```
single<IRandomNumberRepository> { RandomNumberRepository() }
factory { NumberUseCase(get()) }
```
Also, viewModels will be declared here as:
` viewModel { YourViewModel(get()) }`

EntyR committed:
Replaced livedata with flow, changed package name

EntyR committed:
Combined use cases into 1 file, rename repository interface and implementing class, added koin dependency for repository and database

Combined use cases into 1 file, rename repository interface and implementing class, added koin dependency for repository and database

AlexeyKolosov commented on domain/src/main/java/com/harman/roomdbapp/domain/use_cases/RandomNumberUseCase.kt:
```diff
@@ -0,0 +1,17 @@
+        repository.addNumber(number)
+    }
+
+    fun getNumbers(): Flow<List<RandomNumber>> {
```
shorter: 
```
    fun getNumbers() = repository.getNumbers()
```

AlexeyKolosov commented on domain/src/main/java/com/harman/roomdbapp/domain/use_cases/RandomNumberUseCase.kt:
```diff
@@ -0,0 +1,17 @@
+
+class RandomNumberUseCase(private val repository: IRandomNumberRepository) {
+
+    suspend fun addNumber(number: RandomNumber){
```
shorter: 
```
    suspend fun addNumber(number: RandomNumber) = repository.addNumber(number)
```
